samples/boards/nrf/system_off/src/retained.c

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <string.h>
+#include <zephyr.h>
+#include <devicetree.h>
+#include <sys/byteorder.h>
+#include <sys/crc.h>
+#include <hal/nrf_power.h>
+#include "retained.h"
+
+/* nRF51 and nRF52 RAM (really, RAM AHB "slave"s) are partitioned as:
+ * * Up to 8 blocks of two 4 KiBy byte "small" sections
+ * * A 9th block of with 32 KiBy "large" sections
+ *
+ * At time of writing the maximum number of large sections is 6, all
+ * within the first large block.  Theoretically there could be more
+ * "sections" in the 9th section, and possibly more blocks.
+ */
+
+/* Inclusive address of RAM start */
+#define SRAM_BEGIN (uintptr_t)DT_REG_ADDR(DT_NODELABEL(sram0))
+
+/* Exclusive address of RAM end */
+#define SRAM_END (SRAM_BEGIN + (uintptr_t)DT_REG_SIZE(DT_NODELABEL(sram0)))
+
+/* Size of a controllable RAM section in the small blocks*/
+#define SMALL_SECTION_SIZE 4096
+
+/* Number of controllable RAM sections in each of the lower blocks */
+#define SMALL_SECTIONS_PER_BLOCK 2
+
+/* Span of a small block */
+#define SMALL_BLOCK_SIZE (SMALL_SECTIONS_PER_BLOCK * SMALL_SECTION_SIZE)
+
+/* Number of small blocks */
+#define SMALL_BLOCK_COUNT 8
+
+/* Span of the SRAM area covered by small sections */
+#define SMALL_SECTION_SPAN (SMALL_BLOCK_COUNT		\
+	* SMALL_SECTIONS_PER_BLOCK * SMALL_SECTION_SIZE)
+
+/* Inclusive address of the RAM range covered by large sections */
+#define LARGE_SECTION_BEGIN (SRAM_BEGIN + SMALL_SECTION_SPAN)
+
+/* Size of a controllable RAM section in large blocks */
+#define LARGE_SECTION_SIZE 32768
+
+/* Set or clear RAM retention in SYSTEM_OFF for the provided object.
+ *
+ * @note This only works for nRF52 with the POWER module.  The other
+ * Nordic chips use a different low-level API, which is not currently
+ * used by this function.
+ *
+ * @param ptr pointer to the start of the retainable object
+ *
+ * @param len length of the retainable object
+ *
+ * @param enable true to enable retention, false to clear retention
+ */
+static int ram_range_retain(const void *ptr,
```

Comment:
I wonder if this function should be moved to nrfx.  AFAICT, it has no dependencies on any Zephyr-related functionality at all. Just a suggestion from my side.

Reply:
It does depends on `SRAM_BEGIN` and `SRAM_END` mind you, but that is information that nrfx should have already without Devicetree?

Reply:
I don't know whether nrfx has this; if it does, it wasn't obvious.

And use this function should be provided by nrfx, where it should be extended to also support nRF51, nRF53, and possibly nRF91.  That functionality is not currently present; @anangl has agreed to help push it through inside Nordic.